python/thunder/rdds/series.py

```diff
@@ -636,6 +636,111 @@ def query(self, inds, var='inds', order='F', onebased=True):
+        taken from the key of the first Series record in that partition. If the Series object is already
+        sorted and no records have been removed by filtering, then the resulting output should be equivalent
+        to what one would get from calling myImages.saveAsBinarySeries().
+
+        The json config file written out by this method takes the value of the 'nvalues' key from the
+        length of this object's .index attribute. If the index has been modified to have a different length
+        than the number of values in a Series record, then the resulting data files will be read back in
+        incorrectly.
+
+        Parameters
+        ----------
+        outputdirname : string path or URI to directory to be created
+            Output files will be written underneath outputdirname. This directory must not yet exist
+            (unless overwrite is True), and must be no more than one level beneath an existing directory.
+            It will be created as a result of this call.
+
+        overwrite : bool
+            If true, outputdirname and all its contents will be deleted and recreated as part
+            of this call.
+        """
+        import cStringIO as StringIO
+        import struct
+        from thunder.rdds.imgblocks.blocks import SimpleBlocks
+        from thunder.rdds.fileio.writers import getParallelWriterForPath
+        from thunder.rdds.fileio.seriesloader import writeSeriesConfig
+
+        def partitionToBinarySeries(kvIter):
+            """Collects all Series records in a partition into a single binary series record.
+            """
+            keypacker = None
+            firstKey = None
+            buf = StringIO.StringIO()
+            for seriesKey, series in kvIter:
+                if keypacker is None:
+                    keypacker = struct.Struct('h'*len(seriesKey))
+                    firstKey = seriesKey
+                # print >> sys.stderr, seriesKey, series, series.tostring().encode('hex')
+                buf.write(keypacker.pack(*seriesKey))
+                buf.write(series.tostring())
+            val = buf.getvalue()
+            buf.close()
+            # we might have an empty partition, in which case firstKey will still be None
+            if firstKey is None:
+                return iter([])
+            else:
+                label = SimpleBlocks.getBinarySeriesNameForKey(firstKey) + ".bin"
+                return iter([(label, val)])
+
+        writer = getParallelWriterForPath(outputdirname)(outputdirname, overwrite=overwrite)
+
+        binseriesrdd = self.rdd.mapPartitions(partitionToBinarySeries)
+
+        binseriesrdd.foreach(writer.writerFcn)
+        # note that we use len(self.index) here for nvalues! we don't really have a great alternative
+        # to this at the moment, short of calling first(). 
+        writeSeriesConfig(outputdirname, len(self.dims), len(self.index), dims=self.dims.count,
```

Comment:
Agree on getting `nkeys` from a first call, and also using the first call to get the `nrecords`. So that's a great change all around.

Also agree in general about standardizing the Series file format some more (see a comment on this below). But with regards to `dims`, I'm pretty sure currently, whether or not `dims` is in the `conf.json` file, it's not used to set that piece of metadata. We may want to change that, and use it to set the metadata if it's present. But the key point is that `dims` is not strictly required to load Series data (in contrast to, say, `nvalues`). So I don't see this as making the metadata itself optional, it's about storing by default only what's required to load it in, and storing other things optionally to provide useful information (and maybe avoid subsequent operations post-loading).

We can continue this discussion offline =)